### PR TITLE
Mod data content for Gamebryo

### PR DESCRIPTION
**Pull request by Holt59**
Implementation of `ModDataContent` feature for Gamebryo:

- Code mostly from the original `getContents` in `ModInfo` (`modorganizer` project).
- The `GamebryoModDataContent` can be inherited to disable existing content types or add new one:
  - I wanted to avoid having to copy the logic of the two methods in most game plugins, so I came up with this solution.
- The feature is registered in the Gamebryo plugin itself, unlike other features. Game that uses the plugin can erase the feature by registering another `ModDataContent`:
  - I had to `delete` the previously registered feature in `registerFeature`. Also note that currently, when changing the game in MO2, features are not freed, might be something to clean.

### CODE REVIEW

**Comment by Al12rs** [comment]
I'm not sure abut gamebryo registering the feature, what was wrong with having the main logic be in protected methods and the default implementation be just calling those methods? Or was it to not have to edit all the other plugins at all?

**Comment by Al12rs** [comment]
If we need to add support for another bethesda game like fallout 76 or tesVI that need a new content, how should that be added? In gamebryo and then change all other game plugins to disable it?

**Comment by Holt59** [comment]
> I'm not sure abut gamebryo registering the feature, what was wrong with having the main logic be in protected methods and the default implementation be just calling those methods? Or was it to not have to edit all the other plugins at all?

It was mostly to a void having to edit all the game plugins, but in the end I think that only Skyrim and Enderal use the default. Skyrim SE and Skyrim VR should not have the "SkyrProc Patcher" content, and FO4/FO4 VR also need some slight modifications.

> If we need to add support for another bethesda game like fallout 76 or tesVI that need a new content, how should that be added? In gamebryo and then change all other game plugins to disable it?

No, you can add extra ones:
```cpp
class Fallout4ModDataContent: GamebryoModDataContent {
protected:
  enum Fallout4Content {
    CONTENT_MATERIAL = CONTENT_MODGROUP + 1
  };

public:
  Fallout4ModDataContent(GameGamebryo const* gamePlugin) : 
    GamebryoModDataContent(gamePlugin)
  {
    m_Enabled[CONTENT_SKYPROC] = false;
  }

  std::vector<Content> getAllContents() const override 
  {
    auto contents = GamebryoModDataContent::getAllContents();
    contents.push_back(
       Content(CONTENT_MATERIAL, "Materials", ":/MO/gui/content/material"))
    return contents;
  }

  std::vector<int> getContentsFor(
    std::shared_ptr<const MOBase::IFileTree> fileTree) const override 
  {
    auto contents = GamebryoModDataContent::getContentsFor(fileTree);
    for (auto e: *fileTree) {
      if (e->compare("materials") == 0) {
        contents.push_back(CONTENT_MATERIAL);
        break; // Early break if you have nothing else to check.
      }
    }
    return contents;
  }
};
```

**Comment by Al12rs** [comment]
Then maybe use a CONTENT_LAST_ITEM value that can be used to add items to gamebryo without having to keep CONTENT_MODGROUP as the last, this way it's more apparent to maintainers that they need to update that value and order doesn't really matter.

**Comment by Al12rs** [comment]
If you need to edit the other plugins as well I would just let them register the feature like all the other features are currently handled.

File: src/gamebryo/gamebryomoddatacontent.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#include <scriptextender.h>
+
+#include "gamegamebryo.h"
+
+GamebryoModDataContent::GamebryoModDataContent(GameGamebryo const* gamePlugin) : 
+  m_GamePlugin(gamePlugin), m_Enabled(CONTENT_MODGROUP + 1, true) { }
+
+std::vector<GamebryoModDataContent::Content> GamebryoModDataContent::getAllContents() const {
+  static std::vector<Content> GAMEBRYO_CONTENTS{
+    {CONTENT_PLUGIN,    QT_TR_NOOP("Game Plugins (ESP/ESM/ESL)"), ":/MO/gui/content/plugin"},
```
Comment (Al12rs) [inline src/gamebryo/gamebryomoddatacontent.cpp]: remove the ESP/ESM/ESL as it's too long for filters view.

Comment (Holt59) [reply src/gamebryo/gamebryomoddatacontent.cpp]: Maybe remove the "Game" instead? `Plugins (ESP/ESM/ESL)`?

Comment (Al12rs) [reply src/gamebryo/gamebryomoddatacontent.cpp]: agreed

**Review by Al12rs — APPROVED** [review]

File: src/gamebryo/gamegamebryo.h [hunk 2]
```diff
@@ -128,6 +128,10 @@ class GameGamebryo : public MOBase::IPluginGame,
 
   template <typename T>
   void registerFeature(T *type) {
+    auto index = std::type_index(typeid(T));
```
Comment (isanae) [inline src/gamebryo/gamegamebryo.h:131]: `index` can be reused below

File: src/gamebryo/gamebryomoddatacontent.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @brief ModDataContent for GameBryo games.
+ *
+ */
+class GamebryoModDataContent : public ModDataContent {
+protected:
+
+  /**
+   * Note: These are used to index m_Contents so should have standard
```
Comment (isanae) [inline src/gamebryo/gamebryomoddatacontent.h]: Should be `m_Enabled`, not `m_Contents`, I think.

File: src/gamebryo/gamebryomoddatacontent.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+   * @return the IDs of the content in the given tree.
+   */
+  virtual std::vector<int> getContentsFor(std::shared_ptr<const MOBase::IFileTree> fileTree) const override;
+
+protected:
+
+  GameGamebryo const* const m_GamePlugin;
+
+  // List of enabled contents:
+  std::vector<bool> m_Enabled;
```
Comment (isanae) [inline src/gamebryo/gamebryomoddatacontent.h:67]: Is this vector strictly used for `EContent`, or do you expect game plugins that add new contents to also use it? If it's the former, then it should probably be private.

File: src/gamebryo/gamebryomoddatacontent.h [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+  /**
+   *
+   */
+  GamebryoModDataContent(GameGamebryo const* gamePlugin);
+
+  /**
+   * @return the list of all possible contents for the corresponding game.
+   */
+  virtual std::vector<Content> getAllContents() const override;
```
Comment (isanae) [inline src/gamebryo/gamebryomoddatacontent.h:51]: This is somewhat of a personal choice, but `virtual` is redundant. I prefer `virtual` in the base class, and `override` in the derived classes. YMMV.

File: src/gamebryo/gamebryomoddatacontent.h (hunk 4 above)
Comment (Holt59) [reply src/gamebryo/gamebryomoddatacontent.h:67]: Plugins that disable contents use it, e.g., the morrowind game plugin (well actually, all except the Skyrim/Enderal I think... ).

Comment (isanae) [reply src/gamebryo/gamebryomoddatacontent.h:67]: Right, I saw that later. I prefer protected member functions instead of data, but I don't care much.

**Pull request merged**